### PR TITLE
#1580 Create remote sketch

### LINKED ISSUE #1580: Allow creation of sketches in Remote Sketchbook
### Describe the request

Allow the user to create new sketches in their "**Remote Sketchbook**" via the Arduino IDE.

🙂 This gives the user the ability to create new sketches via whichever method is most convenient to them at the time.

### Describe the current behavior

The sketchbook is a convenient place to store Arduino sketches. The sketches in the sketchbook folder are easily accessible via the "**Sketchbook**" view and **File > Sketchbook** menu in the Arduino IDE.

In addition to the traditional sketchbook which exists on the user's local drive, Arduino IDE has [a "**Remote Sketchbook**" linked to Arduino Cloud](https://docs.arduino.cc/software/ide-v2/tutorials/ide-v2-cloud-sketch-sync).

The sketches in the Remote Sketchbook are pulled from Arduino Cloud. They can then be edited in Arduino IDE and the changes pushed back to Arduino Cloud. However, it is not possible to create sketches in the Remote Sketchbook via Arduino IDE. New sketches must always be created on Arduino Cloud.

🙁 Users may find it inconvenient to create new sketches via Arduino Cloud during periods when they are otherwise doing all sketch development work using Arduino IDE.

### Arduino IDE version

2.0.1-snapshot-2f5afe0

### Operating system

All

### Operating system version

Any

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### PR DESCRIPTION
**Pull request by kittaakos**
### Motivation

To create a remote sketch from IDE2 without opening https://create.arduino.cc/editor in a browser.

### Change description

 - `File` > `New` has been renamed to `File` > `New Sketch`.
 - New menu item to create a remote sketch from IDE2: `File` > `New Remote Sketch`.
    Screen Shot 2022-10-31 at 16 09 17
    - If not signed into Arduino Cloud, the menu item is disabled.
        Screen Shot 2022-10-31 at 16 13 47
    - If the _Remote Sketchbook_ is hidden, the menu item is not visible. (`Advanced` > `Show/Hide Remote Sketchbook` or `"arduino.cloud.enabled": false` in `settings.json`)
        Screen Shot 2022-10-31 at 16 17 44
 - Can create a new sketch from the sketchbook.

    https://user-images.githubusercontent.com/1405703/199044937-c482cb79-d827-4ecc-8e79-371a97b568fc.mp4
 - When logged in to Arduino Cloud and the _Remote Sketchbook_ is not hidden, can create a new remote sketch from the Sketchbook.
 
    https://user-images.githubusercontent.com/1405703/199046522-ba14d353-b954-4237-9135-032577720fee.mp4
 - Remote sketch creation gracefully handles conflicts (HTTP 409)

    https://user-images.githubusercontent.com/1405703/199047963-d63fe099-6e74-4053-b583-f22bacb9ec0f.mp4

 - ~After creating the new remote sketch, IDE2 asks if the sketch has to be pulled and opened in a new window.~
 
    

   - ~`Yes` will pull the remote sketch content and open the sketch in a new window.~
   - ~Select `Always` if you want to pull the remote sketch automatically and open it in a new window.~
   - ~Select `Never` to never do anything automatically after the remote sketch creation. Don't worry, you can still pull your sketch manually and open it.~
   - ~You can re-configure the behavior of the notification by tuning the new `arduino.cloud.sketchOpenStrategy` [advanced setting](https://github.com/arduino/arduino-ide/blob/main/docs/advanced-usage.md). The possible values are `"Ask"`, `"Never"`, and `"Always"`. The default is `"Ask"`.~
        
 - After creating the new remote sketch, IDE2 pulls and opens it in a new window.
 - After the remote sketch creation, IDE2 pulls the content, opens the new window, and reveals the sketch in the _Remote Sketchbook_ widget.

    https://user-images.githubusercontent.com/1405703/199052586-08b28a54-3241-4387-8072-991a183fa498.mp4
 - Handles gracefully when there is a mid-air deletion before the pull and open.

    https://user-images.githubusercontent.com/1405703/199054133-da5b920c-6bcd-49da-9a8c-ca969067fced.mp4

### Other information

Closes #1580

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts [hunk 1]
```diff
@@ -149,4 +151,119 @@ export class ElectronMainMenuFactory extends TheiaElectronMainMenuFactory {
     }
     return [];
   }
+
+  // Copied from 1.25.0 Theia as is to customize the enablement of the menu items.
+  // Source: https://github.com/eclipse-theia/theia/blob/ca417a31e402bd35717d3314bf6254049d1dae44/packages/core/src/electron-browser/menu/electron-main-menu-factory.ts#L125-L220
+  // See https://github.com/arduino/arduino-ide/issues/1533
```
Comment (AlbyIanna) [inline arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts:157]: Cool!
This means we can also remove the workaround with `PlaceholderMenuNode` we applied in [cloud-sketchbook-contributions.ts](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/widgets/cloud-sketchbook/cloud-sketchbook-contributions.ts), [interface-scale.ts](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/contributions/interface-scale.ts), [sketch-control.ts](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/contributions/sketch-control.ts), [user-fields.ts](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/contributions/user-fields.ts) and [user-fields.ts](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/contributions/user-fields.ts), am I right?

(I'm not suggesting to necessarily do it in this PR, I just want to keep it in mind)

Comment (kittaakos) [reply arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts:157]: Yes. Thanks for noticing

**Review by AlbyIanna — CHANGES_REQUESTED** [review]
When creating a new remote sketch with multiple IDE windows open, only the sketchbook in the current window is updated to show the newly created item. I need to perform a manual refresh to see the change:

https://user-images.githubusercontent.com/6939054/199768198-6d1cb6d2-f9c5-496c-9a2a-284cd91c5c6d.mov


This doesn't match the behaviour of the local sketchbook, which shows the new sketch in all the windows as soon as it is saved:

https://user-images.githubusercontent.com/6939054/199768895-138720c6-94a3-4794-af3a-6d48b1950abd.mov

**Comment by kittaakos** [comment]
> When creating a new remote sketch with multiple IDE windows open, only the sketchbook in the current window is updated to show the newly created item. I need to perform a manual refresh to see the change:

Yes, but it's the same as with 2.0.1, if you sync one window, you do not sync the other window.


https://user-images.githubusercontent.com/1405703/199770820-d3f642bc-5533-44a6-a627-899600ca4213.mp4


> behaviour of the local sketchbook

IDE2 can poll, but we did not plan the remote sketches with polling on purpose. Users can sync.

**Comment by AlbyIanna** [comment]
> IDE2 can poll, but we did not plan the remote sketches with polling on purpose. Users can sync.

I wasn't thinking of a polling, I thought we could make use of an event to tell every client that they should update the tree. To me it makes sense because we are aware of when a remote sketch is created from the IDE2, so we could easily update every client. I'm not suggesting to update the tree also when a remote sketch is created from the webide. @kittaakos what do you think about that?

**Comment by kittaakos** [comment]
> so we could easily update every client.

Are you aware that IDE2 generates the new remote sketch on the client?

**Comment by AlbyIanna** [comment]
> Are you aware that IDE2 generates the new remote sketch on the client?

I am. In order to do that we should either interact with the backend (which means creating a new service I guess) or with electron-main (which probably means making use of IPC messages). 

> so we could easily update every client

Maybe I used the word "easily" a little lightly here, since of course it's not just as easy as listening to a simple event. But still it shouldn't be so complicated right?

Anyway, to me we could also add this as an improvement later on if you prefer it, as it is outside the scope of this PR.

**Comment by kittaakos** [comment]
> 1. to add a shortcut to for **New Remote Sketch**

👍 I do not have a strong preference for this. We can do it.

> 2\. to add **New Sketch** and **New Remote Sketch** to the Command Palette, so that can be triggered with `ctrl/cmd + shift + P` and then searching for a substring, like this:

Probably no. IDE2 never advertised its features via the _Command Palette_, and I do not want to open Pandora's box in the context of this PR. Please open a follow-up if you're interested in implementing it. Just because IDE2 can have yet another way to create a sketch, it does not necessarily have to have it. One can create a sketch from a keybinding, the sketchbook, and the `File` menu. I think these are sufficient.

**Comment by kittaakos** [comment]
We do the followings:
 - IDE2 does not prompt but always automatically pulls and opens the new remote sketch.
 - When creating a new remote sketch from one window, the second window must be updated with the new, pulled remote sketch.
 - Add keybinding to create a new remote sketch. It should be <kbd>Ctrl/⌘</kbd>+<kbd>Alt</kbd>+<kbd>N</kbd>.
 - IDE2 features won't be exposed as commands in the context of this PR. The follow-up task is #1637.
 - IDE2 does not yet show the [`cloud`](https://github.com/microsoft/vscode-codicons/blob/main/src/icons/cloud.svg) in the title bar if a remote sketch is opened. A follow-up task is tracked under #1638.

**Comment by kittaakos** [comment]
> This doesn't match the behaviour of the local sketchbook, which shows the new sketch in all the windows as soon as it is saved:

It does not work for me.


https://user-images.githubusercontent.com/1405703/199941810-44821979-f6b2-4437-89ba-78a6b1a8f27a.mp4

**Comment by kittaakos** [comment]
> * When creating a new remote sketch from one window, the second window must be updated with the new, pulled remote sketch.

#444

**Comment by kittaakos** [comment]
Please review. Thank you!

**Comment by 91volt** [comment]
@kittaakos 
Thank you, amazing work. 

Testing it in production, the only issue I see is the missing user feedback after inserting the cloud sketch name and before the sketch is open. After naming, takes around 14 secs for the sketch to get pulled, and ~20s to open the workspace (video); a solution could be an extra step in the dialog after inserting the name, showing a spinner while the sketch get created and pulled, similarly to what we do in the firmware updater (screenshot)

@davegarthsimpson @ubidefeo @AlbyIanna @kittaakos 
Do you agree on that? And in case, could we address this with another issue and consider this PR approved for its fundamental scope?

https://user-images.githubusercontent.com/92443068/200873414-33af5188-3ff7-4e21-86af-224f5f4becb3.mov

Schermata 2022-11-09 alle 16 00 50

**Comment by 91volt** [comment]
> Why block the UI? Why does IDE2 not show a progress notification in the right corner? Showing definite or indefinite progress this way is known by the users: it works for lib and platform installation/uninstallation and verify/upload. Can we stick to that?

Because the user is going to be dragged out of focus as soon as the new workspace appears, and it takes about 20s: is a bit annoying if we consider that, the user start writing a line of code, and as soon the sketch get pulled is dragged out of focus.
In any case your proposal is a viable solution and we can discuss it with @ubidefeo, but as I said before:

> could we address this with another issue and consider this PR approved for its fundamental scope?

File: arduino-ide-extension/src/browser/widgets/cloud-sketchbook/cloud-sketchbook-tree-model.ts [hunk 2]
```diff
@@ -81,6 +77,50 @@ export class CloudSketchbookTreeModel extends SketchbookTreeModel {
...
+        }
+      }
+    }
+  }
+
+  protected override isRootAffected(changes: FileChangesEvent): boolean {
+    return changes.changes
+      .map(({ resource }) => resource)
+      .some(
+        (uri) => uri.parent.toString().startsWith(LocalCacheUri.root.toString()) // all files under the root might effect the tree
```
Comment (AlbyIanna) [inline arduino-ide-extension/src/browser/widgets/cloud-sketchbook/cloud-sketchbook-tree-model.ts:110]:
```suggestion
        (uri) => uri.parent.toString().startsWith(LocalCacheUri.root.toString()) // all files under the root might affect the tree
```

**Comment by AlbyIanna** [comment]
> a solution could be an extra step in the dialog after inserting the name, showing a spinner while the sketch get created and pulled, similarly to what we do in the firmware updater (screenshot)
Do you agree on that? And in case, could we address this with another issue and consider this PR approved for its fundamental scope?

I'd like to avoid blocking the UI if possible. Maybe we can handle it in a non-blocking notification with a loading indicator or something like that. But anyway, I agree that we can discuss this separately. This PR looks already good for me.

**Review by AlbyIanna — APPROVED** [review]
I tested it again and it works great for me. Also code is good of course.

Thank you @kittaakos for addressing my remarks!

**Review by 91volt — APPROVED** [review]

**Pull request merged**